rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -98,17 +100,77 @@ def __str__(self) -> str:
         )
 
 
-@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
-class CostBasisEvents:
-    used_acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    spends: List[AssetSpendEvent] = field(init=False)
+class BaseAcquisitionsOrder(metaclass=ABCMeta):
+    _acquisitions: deque[AssetAcquisitionEvent]
 
-    def __post_init__(self) -> None:
-        """Using this since can't use mutable default arguments"""
-        self.used_acquisitions = []
-        self.acquisitions = []
+    def __init__(self) -> None:
+        self._acquisitions = deque()
+
+    @abstractmethod
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """The core method. Should be implemented by subclasses.
+        This method takes a new acquisition and decides where to insert it
+        and thus determines the PnL order.
+        """
+        ...
+
+    def processing_iterator(self) -> Iterator[AssetAcquisitionEvent]:
+        """Iteration method over acquisition events.
+        We can't return here Tuple of AssetAcquisitionEvents as we need to return
+        the first event each time but _acquisitions may be not modified between iterations.
+        """
+        while len(self._acquisitions) > 0:
+            yield self._acquisitions[0]
+
+    def get_acquisitions(self) -> Tuple[AssetAcquisitionEvent, ...]:
+        """Returns read-only _acquisitions"""
+        return tuple(self._acquisitions)
+
+    def consume_result(self, used_amount: FVal) -> None:
+        """This function should be used to consume results of the
+        currently processed event (received from __next__)
+        The current event's remaining_amount will be decreased by used_amount
+        If event's remaining_amount will become ZERO, the event will be deleted
+        May raise:
+        - IndexError if the method was called when acquisitions were empty
+        """
+        assert ZERO <= used_amount <= self._acquisitions[0].remaining_amount, \
+            f'Used amount must be in the interval [0, {self._acquisitions[0].remaining_amount}] but it was {used_amount}'  # noqa: E501p
```

Comment:
```suggestion
        assert ZERO <= used_amount <= self._acquisitions[0].remaining_amount, \
            f'Used amount must be in the interval [0, {self._acquisitions[0].remaining_amount}] but it was {used_amount}'  # noqa: E501
```

Reply:
These kind of assertions are generally not needed as I said before.

We can keep it for a bit to see if it ever hits in the short term, but we don't assert on things that can't happen in methods in general.

We can leave it as a comment so the devs know what to expect though.

Reply:
Alright, added a comment